### PR TITLE
ENH: `stats.variation` empty input handling and tests; doc improvement

### LINKED ISSUE #592: Statistics Review: variation (Trac #65)
_Original ticket http://projects.scipy.org/scipy/ticket/65 on 2006-04-02 by @rkern, assigned to @rkern._

The function `variation` in file source:trunk/Lib/stats/stats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### PR DESCRIPTION
**Pull request by aeklant**
closes gh-592

### CODE REVIEW

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -1088,14 +1091,36 @@ def variation(a, axis=0):
...
+    >>> from scipy import stats
+    >>> a = np.linspace(stats.erlang.ppf(0.01, 1), 100)
+    >>> stats.variation(a)
+    0.58889670528253069
+
     """
     a, axis = _chk_asarray(a, axis)
+
+    if a.size == 0:
+        return np.nan
```
Comment (josef-pkt) [inline scipy/stats/stats.py]: Just thought about it.
For functions that have an axis keyword, there will be possibly a multidimensional answer. If/When we add `missing`, then there could be some elements that are missing/nan similar to the current `mstats`.

In the same way, the user could give us an array that is empty, `a.size == 0` but has more than one dimension, rows or columns. I vaguely remember similar discussions about the empty behavior on the numpy mailing list.

e.g. variation(np.array([]).reshape(0, 0, 0)

```
>>> np.mean(np.array([]).reshape(0,0,0), axis=0)
array([], shape=(0, 0), dtype=float64)
C:\programs\WinPython-64bit-3.4.3.1\python-3.4.3.amd64\lib\site-packages\numpy\core\_methods.py:59: RuntimeWarning: Mean of empty slice.
  warnings.warn("Mean of empty slice.", RuntimeWarning)
```

Comment (rgommers) [inline scipy/stats/stats.py]: Technically correct I think, but an extreme corner case. Let's think about if we can fix this without a lot of repetitive boilerplate code. There are already functions that do this, but not very consistently probably. Example from ``scoreatpercentile``:

    if a.size == 0:
        # empty array, return nan(s) with shape matching `per`
        if np.isscalar(per):
            return np.nan
        else:
            return np.ones(np.asarray(per).shape, dtype=np.float64) * np.nan

Comment (rgommers) [inline scipy/stats/stats.py]: Actually, currently the correct shape is returned:

    In [5]: x = np.array([]).reshape(0, 5)

    In [7]: stats.variation(x)
    /home/rgommers/.local/lib/python2.7/site-packages/numpy/core/_methods.py:115: RuntimeWarning: invalid value encountered in true_divide
      ret, rcount, out=ret, casting='unsafe', subok=False)
    /home/rgommers/.local/lib/python2.7/site-packages/numpy/core/_methods.py:69: RuntimeWarning: invalid value encountered in true_divide
      ret, rcount, out=ret, casting='unsafe', subok=False)
    Out[7]: array([ nan,  nan,  nan,  nan,  nan])

Comment (rgommers) [inline scipy/stats/stats.py]: The logic is here: https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/reduction.c#L194

I'm not aware of a public function in numpy that provides the right shape (I could be missing it - anyone?), but it's not too difficult to reimplement here in a utility function.

**Comment by aeklant** [comment]
This seems like an issue that will probably span throughout a large portion of `scipy.stats`. It also seems like the correct shape is returned (at least here, in `stats.variation`), albeit with some warnings.

Could we agree on ignoring it for now in favor of the remaining items of the `scipy.stats` improvement project and then circling back to it if time allows for it?

**Comment by rgommers** [comment]
@aeklant it's a bit difficult to ignore, because when merging this PR we break something that is now correct. It's fine to pick it up later (I agree there are more important things to worry about), but it means that this PR will have to stay open for a while.

**Comment by WarrenWeckesser** [comment]
@aeklant, it looks like this PR stalled a few years ago.  Are you interested in resuming work on this?

As @josef-pkt and @rgommers noted back in 2015, the handling of `a.size == 0` in the PR is not correct.  We must handle the `axis` argument consistently, even when `a.size` is 0.  A good model to follow is the behavior of `numpy.mean`.

For example,

```
In [65]: z = np.empty((3, 0))

In [66]: z.shape
Out[66]: (3, 0)
```

Note that the shape of the following result is (0,), and no warning is generated.  That is how `variation` should behave also.

```
In [67]: z.mean(axis=0)
Out[67]: array([], dtype=float64)
```

The follow returns an array with shape `(3,)` that is all `nan`s. It generates a warning because it is attempting to take the mean of empty arrays.  `variation` should behave the same.

```
In [68]: z.mean(axis=1)
/Users/warren/a2019.10/bin/ipython:1: RuntimeWarning: Mean of empty slice.
  #!/Users/warren/a2019.10/bin/python
/Users/warren/a2019.10/lib/python3.7/site-packages/numpy/core/_methods.py:154: RuntimeWarning: invalid value encountered in true_divide
  ret, rcount, out=ret, casting='unsafe', subok=False)
Out[68]: array([nan, nan, nan])
```

**Comment by WarrenWeckesser** [comment]
I should have checked before commenting above:  the current implementation is actually pretty close to what we want, because in the basic case it ends up calling `a.std(axis) / a.mean(axis)`, and those methods do the right thing in most cases:

```
In [6]: import scipy

In [7]: scipy.__version__
Out[7]: '1.5.0.dev0+eefd95f'

In [8]: from scipy.stats import variation

In [9]: z = np.empty((3, 0))

In [10]: variation(z, axis=0)
Out[10]: array([], dtype=float64)

In [11]: variation(z, axis=1)
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:217: RuntimeWarning: Degrees of freedom <= 0 for slice
  keepdims=keepdims)
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:186: RuntimeWarning: invalid value encountered in true_divide
  arrmean, rcount, out=arrmean, casting='unsafe', subok=False)
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:207: RuntimeWarning: invalid value encountered in true_divide
  ret, rcount, out=ret, casting='unsafe', subok=False)
/Users/warren/mc37scipy/lib/python3.7/site-packages/scipy-1.5.0.dev0+eefd95f-py3.7-macosx-10.7-x86_64.egg/scipy/stats/stats.py:1038: RuntimeWarning: Mean of empty slice.
  return a.std(axis) / a.mean(axis)
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:154: RuntimeWarning: invalid value encountered in true_divide
  ret, rcount, out=ret, casting='unsafe', subok=False)
Out[11]: array([nan, nan, nan])
```
The shapes and contents of the return values are correct, but there are too many incidental warnings in the second case.

**Comment by WarrenWeckesser** [comment]
It looks like the code and the docstring for `variation` have been updated since this PR was created, making the PR mostly obsolete, so I'm closing it.  @aeklant, let us know if you don't agree and would like to contribute additional changes.

**Pull request closed without merging**